### DISCUSSION TITLE
How to set Kafka log.retention.hours with Operator?

### DISCUSSION THREAD

**Discussion by mrocheleau** [strimzi/strimzi-kafka-operator#9812] [General]
Hi, we have a Strimzi jaeger kafka cluster in k8s, version 3.4.0 it looks like.  Our lead kafka box is running low on disk on it's pvc where everything gets stored.

In the past, having noted that the `log.retention.hours` was showing as `168` on startup, I pruned that down by editing the `kafka` resource and adding:
```
  kafka:
    config:
      log.retention.hours: 96
```
To reduce retention from 7 days to 4.  When I did this previously the disk was almost entirely full, and editing and saving that change initiated a restart of the pods and they began to prune down on their own.  

Now some months later we're hit with the same situation but I see the default `168` hours set in again, and traces go back 7 full days.  When I repeat the same steps and edit the `kafka` resource, adding the same `log.retention.hours: 96` configuration line, which wasn't present when I checked this time, nothing happens.  No restarts, no pruning, no log activity on the pods at all.

I'm unsure if any other changes occurred on the setup since the first time some months ago, but I doubt it.  When I check the `/tmp/strimzi.properties` file on the pods it doesn't have anything set for `log.retention.hours` so it must be using defaults...but why when I have it set?

How does one set this value to less than the default, correctly, if not via the `kafka` resource?

**Comment by scholzj** [comment]
I guess you would need to check the logs to see why is not rolling the brokers. You might find useful information in the `.status` section of the KAfka CR and in the Cluster Operator logs.

**Reply by mrocheleau** [reply]
Oh interesting, in the `kafka` CR the status shows:
```
  - lastTransitionTime: "2024-02-14T13:07:14.262472928Z"
    message: The desired Kafka storage configuration contains changes which are not
      allowed. As a result, all storage changes will be ignored. Use DEBUG level logging
      for more information about the detected changes.
    reason: KafkaStorage
    status: "True"
    type: Warning
  - lastTransitionTime: "2024-02-14T13:07:14.262509406Z"
    message: log.message.format.version does not match the Kafka cluster version,
      which suggests that an upgrade is incomplete.
    reason: KafkaLogMessageFormatVersion
    status: "True"
    type: Warning
```

How does one enable debug logging - and would I just restart the pods one by one after or can I initiate a rolling restart via the kafka CR?

**Reply by scholzj** [reply]
These are just warnings, they should not block any rolling updates. You can configure the operator logging through its config map.

**Reply by mrocheleau** [reply]
I manually rolled the kafka pods though after I put in the change to the kafka CR to add the log retention change, that had no effect they came up with the default log retention again.  

I don't see any configmaps for the operator itself, the two with `operator` in the name say don't modify this it's auto generated:
```
NAME                                            DATA   AGE
jaeger-streaming-entity-topic-operator-config   1      325d
jaeger-streaming-entity-user-operator-config    1      325d
jaeger-streaming-kafka-0                        3      325d
jaeger-streaming-kafka-1                        3      325d
jaeger-streaming-kafka-2                        3      325d
jaeger-streaming-sampling-configuration         1      325d
jaeger-streaming-ui-configuration               1      325d
jaeger-streaming-zookeeper-config               2      325d
```

**Reply by mrocheleau** [reply]
And notably from the Warning there, the wording `As a result, all storage changes will be ignored`, not sure if that relates here (is it a storage change?).

**Reply by scholzj** [reply]
You cannot just randomly delete a pod and expect it to fix everything. You need to figure out why it doesn't work first. You can also share the full custom resource and the full log from the operator and someone might try to help you by looking into it.

Storage change means you did some illegal change to `.spec.kafka.storage`.

**Reply by mrocheleau** [reply]
I made no changes to that tiny block, no.  K8s upgrade happened on Feb 14th 2024, but otherwise no changes made in this resource outside of having added the log retention config months ago which since vanished somehow.

Previously we had the disks set to 100GB and completely ran out of space, so I bumped it to 125GB to being it back to life (forget exactly where this was done now) but the field for `.spec.kafka.storage.volumes.size` was set to `100Gi` still.  Updating that to `125Gi` and saving had no effect at all.  No operator logs, no kafka box logs related to it, no restarts, no change to the kafka CR status fields.

There's only a few other options in that storage block, is there a way to know what the defaults are to confirm what the few other items are?  

At present it only has:
```
    storage:
      type: jbod
      volumes:
      - deleteClaim: false
        id: 0
        size: 125Gi
        type: persistent-claim
```

**Reply by scholzj** [reply]
Kubernetes does not support shrinking volume sizes. So if you increased it to 125Gi, you cannot shrink it back to 100Gi. So that is likely what that warning is about.

**Reply by mrocheleau** [reply]
I realize that, yes.  The thing is, looking back at my notes, this kafka CR was where I adjusted the disk size in the first place.  So I upped it by 25GB here months ago, then afterwards set the log retention reduction.  Now, with no changes recorded in our code since then to this stuff, we have lost both of those changes in this CR.  How could that happen?  Is there something higher up the chain feeding in default values?

As I said, I updated it to 125Gi in the `kafka` CR just now and it had no impact.  Is there a way to trigger a rolling restart of the pods?  I can't find documentation on any of these sort of useful operations so have come here.

**Reply by scholzj** [reply]
You can use annotation to trigger manual rolling updates. It is in the docs. But again, doing that blindly will not help you when you don't understand the issue. If the operator ignores you retention configuration, the rolling update will likely not set it. If it is _broker_ for some reason, it might ignore the manual rolling update.

**Reply by mrocheleau** [reply]
I'm trying to understand the issue.  I didn't build this, I've inherited it.  To continue using it we need this retention problem fixed.

I'm unsure where to even implement more verbose logging, the configmaps don't have anything for this and all say to not touch them as they're auto generated.  The only changes made were to disk size of the kafka storage and to log retention (and a TLS cert update from ElasticSearch).  Now two of those three changes are no longer reflected.  A month ago we upgraded K8s in place from 1.26 to 1.28, so everything got restarted of course, but to my knowledge that's it.  

Looking for any suggestions for where to look.  Trying to perform a manual restart via annotation doesn't do anything at all:
```
kubectl -n monitoring annotate strimzipodset jaeger-streaming-kafka strimzi.io/manual-rolling-update=true
```
Is our operator broken in some way?  The operator logs are showing nothing useful at all, it's just a constant cycle of:
```
2024-03-11T19:43:33Z	INFO	Determining whether we should enable the Elasticsearch Operator integration
2024-03-11T19:43:33Z	INFO	Determining whether we should enable the Kafka Operator integration
2024-03-11T19:43:33Z	INFO	found the cronjobs api in batch/v1
2024-03-11T19:43:33Z	INFO	found the horizontalpodautoscalers api in autoscaling/v2
2024-03-11T19:43:33Z	INFO	detecting orphaned deployments.
2024-03-11T19:43:38Z	INFO	Determining whether we should enable the Elasticsearch Operator integration
2024-03-11T19:43:38Z	INFO	Determining whether we should enable the Kafka Operator integration
2024-03-11T19:43:38Z	INFO	found the cronjobs api in batch/v1
2024-03-11T19:43:38Z	INFO	found the horizontalpodautoscalers api in autoscaling/v2
2024-03-11T19:43:38Z	INFO	detecting orphaned deployments.
```
Nothing else in the entire log history on that operator pod.

**Reply by mrocheleau** [reply]
In the docs I see references to increase disk space via the storage block and each one shows three unique volumes one per replica, but I only see a "0" id volume in the `kafka` CR:

`.spec.kafka`:
```
    replicas: 3
    storage:
      type: jbod
      volumes:
      - deleteClaim: false
        id: 0
        size: 125Gi
        type: persistent-claim
```
Unless the jbod type means id: 0/1/2 would be three unique disks per replica, instead of simply "3 replicas".

**Reply by scholzj** [reply]
> I'm trying to understand the issue. I didn't build this, I've inherited it. To continue using it we need this retention problem fixed.
>
> I'm unsure where to even implement more verbose logging, the configmaps don't have anything for this and all say to not touch them as they're auto generated. The only changes made were to disk size of the kafka storage and to log retention (and a TLS cert update from ElasticSearch). Now two of those three changes are no longer reflected. A month ago we upgraded K8s in place from 1.26 to 1.28, so everything got restarted of course, but to my knowledge that's it.

I can understand that it is frustrating if you inherited something you don't understand. But I'm afraid that I do not have a magic wand to figure out the issue without the necessary information. So as I suggested, you can share the full custom resource including the status and the full operator log as a start and maybe that will show something. 

If you cannot find the operator, then maybe someone deleted it and it is not running anymore? That would explain the issue. But it is hard to say - I do not know how it was installed or anything. So I cannot tell you where to look for it. It is named _Strimzi Cluster Operator_ and in most cases that is in the pod name.

The documentation has a chapter describing how to change the log levels of the different components including the Cluster operator. But in any case, you need to know where to find the logs first.

> Is our operator broken in some way? The operator logs are showing nothing useful at all, it's just a constant cycle of:

I do not think these are logs from the Strimzi operator that would be running your Kafka cluster.

**Comment by mrocheleau** [comment]
Ultimately looks like somehow the `kafka` custom resource reverted or was applied overtop incorrectly (though if I didn't do this, which I didn't, then it's unlikely anyone else did) and our disk size and log retention changes we made months back were lost from the CR.  

When the kafka pods restarted they became decoupled or similar from the `kafka` CR, as the CR was failing to perform a downgrade of size back to the original disk size as this is unsupported by k8s of course.  This meant any further changes were ignored.

Updating the disk size in the `kafka` CR also did nothing, presumably it was being very picky about not progressing past the initial incorrect downgrade of sizing request.  Updating annotations to force a rolling restart had no impact.  However updating logging levels for kafka specifically in the strimzi operator's configmap, and restarting the operator, did cause the kafka replicas to restart, possibly only due to the pending annotation based restart request though.  When they restarted the warning cleared on the `kafka` CR and the changes stuck, the new pods came up with the log.retention.hours setting in the `/tmp/strimzi.properties` file correctly and pruning down to that level began happening automatically.  

Root cause is still unknown though, we didn't touch the `kafka` CR to my knowledge between the first disk size increase and today, yet it was missing the changes we implemented on that first change, which were 100% there as we pruned down from a high disk size at that time.
